TEES refactored for batch processing

djmilstein opened the pull request:
Refactor the previously monolithic TEES reader and processor to separate these steps:
1) Invoking the TEES reader
2) Extracting the relevant TEES output as strings from the output directory containing much relevant and irrelevant TEES output
3) Processing strings containing the relevant TEES output extracted in step 2) into INDRA statements

This lays the groundwork to running TEES on AWS with the database, in particular facilitating storing the TEES output as strings in the database before processing.

bgyori commented on indra/sources/tees/parse_tees.py:
```diff
@@ -257,14 +256,18 @@ def index_to_sentence(self, index):
     ----------
-    output_dir: str
-        Directory containing the output of the TEES system
+    a1_text: str
```
Add spaces after argument name everywhere, e.g. `a1_text : str`

bgyori commented on indra/resources/default_config.ini:
```diff
@@ -12,6 +12,9 @@ EIDOSPATH =
 SPARSERPATH = 
 
+# Path to TEES
+TEES_PATH = ~/TEES
```
I'd remove your specific local path from the default config and just leave it blank

bgyori commented on indra/sources/tees/tees_api.py:
```diff
@@ -104,6 +116,197 @@ def process_text(text, pmid=None, tees_path=None, python2_path=None):
+    ----------
+    text: str
+        Text from which to extract relationships
+    tees_path: str
```
Since there is a TEES_PATH config now, this argument is redundant, I think, and could be removed in favor of reading from the config.

bgyori commented on indra/sources/tees/processor.py:
```diff
@@ -16,43 +16,43 @@
-        INDRA statements.
+    a1_text: str
+        The TEES a1 output file, with entity information
+    a2_test: str
```
Change a2_test to a2_text

bgyori commented on indra/sources/tees/tees_api.py:
```diff
@@ -104,6 +116,197 @@ def process_text(text, pmid=None, tees_path=None, python2_path=None):
+def run_tees_on_text(text, tees_path, python2_path):
+    """Runs TEES on the given text in a temporary directory and returns a
+    temporary directory with TEES output. The caller should delete this
+    directory when done with it.
```
To make this a bit more clear, add another sentence like this: "This call produces TEES output files but does not process the TEES output into INDRA Statements."

bgyori commented on indra/sources/tees/tees_api.py:
```diff
@@ -104,6 +116,197 @@ def process_text(text, pmid=None, tees_path=None, python2_path=None):
+    return output_tuple
+
+def extract_relevant_tees_output(output_dir):
+    """Extract the text of the a1, a2, and sentence segmentation files from the
```
Why "relevant" in the name? If not essential, just rename to `extract_output` (as in previous comment `tees` is not needed in the name)

bgyori commented on indra/sources/tees/tees_api.py:
```diff
@@ -104,6 +116,197 @@ def process_text(text, pmid=None, tees_path=None, python2_path=None):
+    tp = TEESProcessor(a1_text, a2_text, sentence_segmentations, pmid)
     return tp
+
+def run_tees_on_text(text, tees_path, python2_path):
```
No need to add `tees` in the name here, just call it `run_on_text`

bgyori commented on indra/sources/tees/tees_api.py:
```diff
@@ -104,6 +116,197 @@ def process_text(text, pmid=None, tees_path=None, python2_path=None):
+        tar_file.extractall(path=tmp_dir, members=extract_these)
+
+        # Read the text of the a1 (entities) file
+        with open(os.path.join(tmp_dir, a1_file), 'r') as f:
```
Encoding should probably be made explicit here and all other places, i.e. open as 'rb' and decode after reading depending on encoding used by TEES output files.

bgyori commented on indra/sources/tees/parse_tees.py:
```diff
@@ -85,38 +85,39 @@ def parse_a1(a1_filename):
+        if len(line) == 0:
+            continue
+        tokens = line.rstrip().split('\t')
+        assert(len(tokens) == 3)
```
Raise an exception (either a built-in one or a custom exception class which inherits from a built-in one) instead of assert here and in several other places that follow

bgyori commented on indra/sources/tees/parse_tees.py:
```diff
@@ -257,14 +256,18 @@ def index_to_sentence(self, index):
-def parse_tees_output_directory(output_dir):
-    """Parses the files in the TEES output directory. And returns a networkx
-    graph.
+def parse_tees_output_files(a1_text, a2_text, sentence_segmentations):
```
Here `tees` is not needed in the name. It also looks like these aren't actually files but rather strings so the function could be named `parse_output`.

djmilstein commented:
Feedback for tees_batch to date now incorporated

On Mon, Apr 30, 2018 at 1:32 AM, Benjamin M. Gyori <notifications@github.com
> wrote:

> *@bgyori* commented on this pull request.
> ------------------------------
>
> In indra/sources/tees/parse_tees.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184911953>:
>
> >
>      Parameters
>      ----------
> -    output_dir: str
> -        Directory containing the output of the TEES system
> +    a1_text: str
>
> Add spaces after argument name everywhere, e.g. a1_text : str
> ------------------------------
>
> In indra/resources/default_config.ini
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184911993>:
>
> > @@ -12,6 +12,9 @@ EIDOSPATH =
>  # Path to the Sparser reading system folder
>  SPARSERPATH =
>
> +# Path to TEES
> +TEES_PATH = ~/TEES
>
> I'd remove your specific local path from the default config and just leave
> it blank
> ------------------------------
>
> In indra/sources/tees/tees_api.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184912153>:
>
> >      return tp
> +
> +def run_tees_on_text(text, tees_path, python2_path):
> +    """Runs TEES on the given text in a temporary directory and returns a
> +    temporary directory with TEES output. The caller should delete this
> +    directory when done with it.
> +
> +    Parameters
> +    ----------
> +    text: str
> +        Text from which to extract relationships
> +    tees_path: str
>
> Since there is a TEES_PATH config now, this argument is redundant, I
> think, and could be removed in favor of reading from the config.
> ------------------------------
>
> In indra/sources/tees/processor.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184912264>:
>
> >
>      Parameters
>      ----------
> -    text: str
> -        Plain text from biomedical publications from which to extract
> -        INDRA statements.
> +    a1_text: str
> +        The TEES a1 output file, with entity information
> +    a2_test: str
>
> Change a2_test to a2_text
> ------------------------------
>
> In indra/sources/tees/tees_api.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184912736>:
>
> >      return tp
> +
> +def run_tees_on_text(text, tees_path, python2_path):
> +    """Runs TEES on the given text in a temporary directory and returns a
> +    temporary directory with TEES output. The caller should delete this
> +    directory when done with it.
>
> To make this a bit more clear, add another sentence like this: "This call
> produces TEES output files but does not process the TEES output into INDRA
> Statements."
> ------------------------------
>
> In indra/sources/tees/tees_api.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184912819>:
>
> > +            os.chdir(pwd)  # Change back to previous directory
> +            # print (e.output.decode('utf-8'))
> +            raise e
> +
> +    except BaseException as e:
> +        # If there was an exception, delete the temporary directory and
> +        # pass on the exception
> +        shutil.rmtree(tmp_dir)
> +        raise e
> +    # Return the temporary directory with the TEES output
> +    output_tuple = extract_relevant_tees_output(tmp_dir)
> +    shutil.rmtree(tmp_dir)
> +    return output_tuple
> +
> +def extract_relevant_tees_output(output_dir):
> +    """Extract the text of the a1, a2, and sentence segmentation files from the
>
> Why "relevant" in the name? If not essential, just rename to
> extract_output (as in previous comment tees is not needed in the name)
> ------------------------------
>
> In indra/sources/tees/tees_api.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184912833>:
>
> >      return tp
> +
> +def run_tees_on_text(text, tees_path, python2_path):
>
> No need to add tees in the name here, just call it run_on_text
> ------------------------------
>
> In indra/sources/tees/tees_api.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184912950>:
>
> > +                    a2_file = m.name
> +                else:
> +                    assert(False)
> +
> +        # There should be exactly two files that match these criteria
> +        if len(extract_these) != 2 or a1_file is None or a2_file is None:
> +            raise Exception('We thought there would be one .a1 and one .a2' +
> +                            ' file in the tarball, but we got %d files total' %
> +                            len(extract_these))
> +            return None, None, None
> +
> +        # Extract the files that we decided to extract
> +        tar_file.extractall(path=tmp_dir, members=extract_these)
> +
> +        # Read the text of the a1 (entities) file
> +        with open(os.path.join(tmp_dir, a1_file), 'r') as f:
>
> Encoding should probably be made explicit here and all other places, i.e.
> open as 'rb' and decode after reading depending on encoding used by TEES
> output files.
> ------------------------------
>
> In indra/sources/tees/parse_tees.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184913089>:
>
> > -            assert(len(info_tokens) == 3)
> -            entity_type = info_tokens[0]
> -            first_offset = int(info_tokens[1])
> -            second_offset = int(info_tokens[2])
> -            offsets = (first_offset, second_offset)
> -
> -            entities[identifier] = TEESEntity(
> -                    identifier,
> -                    entity_type,
> -                    entity_name,
> -                    offsets)
> +    for line in a1_text.split('\n'):
> +        if len(line) == 0:
> +            continue
> +        tokens = line.rstrip().split('\t')
> +        assert(len(tokens) == 3)
>
> Raise an exception (either a built-in one or a custom exception class
> which inherits from a built-in one) instead of assert here and in several
> other places that follow
> ------------------------------
>
> In indra/sources/tees/parse_tees.py
> <https://github.com/sorgerlab/indra/pull/512#discussion_r184913210>:
>
> > @@ -257,14 +256,18 @@ def index_to_sentence(self, index):
>              return None
>
>
> -def parse_tees_output_directory(output_dir):
> -    """Parses the files in the TEES output directory. And returns a networkx
> -    graph.
> +def parse_tees_output_files(a1_text, a2_text, sentence_segmentations):
>
> Here tees is not needed in the name. It also looks like these aren't
> actually files but rather strings so the function could be named
> parse_output.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sorgerlab/indra/pull/512#pullrequestreview-116197563>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOWmzW75tDlHp-enjJA3HXv_e8jbl-w0ks5ttqHrgaJpZM4TlS2J>
> .
>
